### ISSUE TITLE
Nested beforeEach runs even if outer beforeEach fails due to async timeout

### ISSUE THREAD

**Issue by john-swallow-cenduit** [jasmine/jasmine#1882]
## Expected Behavior
If an asynchronous beforeEach is failed due to it breaching the Jasmine DEFAULT_TIMEOUT_INTERVAL I would expect the execution to stop and any beforeEach nested inside it to not execute. 

I would expect that a nested beforeEach should only EVER run if the outer beforeEach has completed. 

## Current Behavior
Jasmine carries on running anyway with the completely counter intuitive result that it can end up interleaving the execution of beforeEach and running all required actions completely out of order.

## Possible Solution
Terminate execution when Jasmine decides that the async has breached the timeout.

Alternatively - carry on running - but don't run UNTIL the beforeEach on the outside has finished or the order of execution is changed.

Workaround: 
Setting jasmine.DEFAULT_TIMEOUT_INTERVAL to some suitably large value.

## Suite that reproduces the behavior (for bugs)
A full solution showing the issue can be found at: 
https://github.com/john-swallow-cenduit/BeforeEachbug

Basically this boils down to:

```
describe('root', () => {
  beforeEach(async () =>   
  {
    
    console.log('root: begin');
    // If this sleep exceeds the default timeout the issue
    // will appear.
    await sleep(5000);
    console.log('root: end');
  });

  describe('inner', () => {
    beforeEach(async () => {
      console.log('inner: begin');
      await sleep(1500);
      console.log('inner: end');
    });

    it('inner it', () => {
      expect().nothing();
    });

  });

  function sleep(delayInMilliseconds: number) {
    return new Promise( resolve => setTimeout(resolve, delayInMilliseconds) );
  }
});
```

The sleep is artificial of course and is just there to prove the point.

The code should run and output:
```
root: begin
<-- Delay here (1)
root: end
inner: begin
<-- Delay here (2)
inner: end
``` 

But what actually happens is that (1) breaches the timeout value and you get this instead:
```
root: begin
inner: begin
root: end
inner: end
```

i.e. code in the inner was run BEFORE all code in the outer root had completed. As the root did not complete successfully I would expect inner NOT to be run but it actually does start executing inner.

## Context
I spotted it while debugging my tests. I used a breakpoint (thereby suspending the code and bringing the timeout into focus) and was then extremely confused when it ran all my test code out of order.

A real world example when NOT debugging would be that I am waiting inside the beforeEach for my Angular app to achieve a stable state using await fixture.whenStable() but due to some problem it doesn't do so in time. I am expecting a failure to occur in that case but the failure should not counter indicate elsewhere in my tests - it should clearly come from the beforeEach in which the problem is occuring.

## Your Environment
Jasmine 3.6.0
Chrome 88.0.4324.150 (Official Build) (64-bit)
node v10.15.0
TypeScript 4.1.5
Windows 10 - desktop

**Comment by sgravrock** [comment]
In your first proposed solution:

> Terminate execution when Jasmine decides that the async has breached the timeout.

I'm not sure if you mean execution of the async function or execution of the entire test suite, so I'll answer both.

Jasmine can't terminate execution of the async function. JS runtimes don't provide that functionality, so we have no choice but to let the async function keep running. That means that it'll inevitably interleave with something: if not a nested beforeEach then a spec or a beforeEach or an afterEach in some other context, or even with Jasmine's own housekeeping code. As you've noticed, that's a pretty confusing situation. And writing code that behaved correctly in the face of that kind of random interleaving is very hard.

As for terminating the entire suite, I think there's a case to be made for it. I might have argued for it if I'd been around, and given the full benefit of hindsight, when the current flavor of async specs was implemented. But there's also a case against it. Many async timeouts in Jasmine specs occur because the async function is waiting for something that will never happen. In those cases there's nothing to interleave, since the remainder of the async function will never execute. Letting execution continue means that the user can get feedback from the rest of their test suite without having to first fix the async failure.

I suspect that those benign async failures, where the async function is waiting for something that will never happen, are by far the most common. (I'd estimate that they're at least 99% of async timeouts in my own experience, although e.g. the Protractor crowd would probably come up with very different numbers.) So we're probably better off optimizing for developer productivity in that case by continuing in the case of async timeouts, even if it makes things more confusing in the case where an async function times out but keeps doing things.

If I understand your second proposed solution, it sounds like it would be akin to not having async timeouts at all. I don't think that would be a very popular change. And as you've discovered, you can approximate it by setting jasmine.DEFAULT_TIMEOUT_INTERVAL to a large value.

Sorry, but this is a really hard problem to solve in the general case.

**Comment by john-swallow-cenduit** [comment]
Thanks for responding but I think we got off on the wrong foot and are now tangential to my prime concern. Let me deal with the tanget first:

"I'm not sure if you mean execution of the async function or execution of the entire test suite, so I'll answer both."

I didn't actually mean either. What I meant by "terminate execution" was don't continue running the test you were running - but only that test. I chose my words carelessly, I admit. I.e. you are running the beforeEach() functions to get setup to run an it(). I'm talking about failing that it(). I wasn't talking about terminating the async call itself. I'm quite happy for Jasmine to decide that the tests has timed out and for it to FAIL when this occurs. That's what I would EXPECT to happen and I'm fine with that - in fact I WANT that. In essence what I am doing in this case is a Protractor crowd thing - I'm waiting for the web app to do something and checking the result.

But I think that was a tangent caused by my poor choice of works. The crux of my actual issue comes from a possible interpretation issue with what a nested beforeEach means and what nesting a beforeEach within an outer beforeEach(async()=> {}). You could look at this two ways:

**Interpretation 1**
Regardless of whether the outer is async or not the outer beforeEach must **complete** before the inner beforeEach starts 
So when the outer before yields control - that yields for other work but the other work does NOT include running the inner beforeEach. Under this interpretation the two beforeEach functions always run in sequence. That's what Jasmine MEANS by a nested beforeEach and it internally makes it happen.

**Interpretation 2**
If the outer beforeEach is async and yields its fine to run the inner beforeEach
So if the outer beforeEach yields while it is waiting get on with running the inner one.

I guess it's the difference between Jasmine running one of the two items of pseudocode:

Interpretation 1:
```
**await** outerBeforeEach();
// Wait for outerBeforeEach to finish and either complete or timeout. Only go forward to the next once this has happened. If the call times out then fail the test do NOT go to the next line. 
innerBeforeEach();
```

or

Interpretation 2:
```
outerBeforeEach();
// Run innerBeforeEach without waiting to see what the result of outerBeforeEach was
innerBeforeEach(); 
```

But at the end of the day Jasmine must have made one of these choices. 

I expected interpretation 1 and was surprised to find that despite the FAILURE (due to timeout) of my outer beforeEach() that the inner beforeEach was even running. They perform setup that MUST go in sequence .If the outer times out the inner is completely meaningless and there's just no point. The separation into two nested beforeEach function is because the tests logically structure that way. The outer sets up common test setup for all the tests and the inner then makes some additional setup that apply to only some test cases.

I can't find any documentation spelling this out. So though I can say that (1) is more natural to me you could say that interpretation 2 is the correct one I guess. So it's either not working as it should - or there's a documentation gap to fill (at least I wasn't able to satisfactorily answer my question with what I could find - I could have missed something.

I want to produce a situation where the system behaves like interpretation 1 so I can get a consistent behaviour that makes sense to me. But right now I don't think I have a way of doing so.

If it just doesn't work like that and interpretation 2 is just the way it is then I guess I can understand - but hopefully you can understand WHY I though interpretation 1 was how it SHOULD be working.

I hope that all makes sense.

**Comment by sgravrock** [comment]
Thanks. I think I have a much better understanding now. The first thing I'd try in your case is the `oneFailurePerSpec` option. You can set it by passing `oneFailurePerSpec=true` on the query string or modifying boot.js if you’re using Jasmine Standalone, or by passing `--stop-on-failure=true` to the rpm runner. In your scenario, that will cause the spec and any remaining before functions in the same or nested describes to be skipped. (`afterEach`/`afterAll` functions will still run, which probably isn’t what you want when you’re stopped on a breakpoint, but I think it’s the best choice overall.)


To clarify, neither of the two interpretations you outlined above matches what Jasmine does. It does a third thing: It waits for each user function to either complete or time out, and then it moves onto the next function regardless of whether the first one produced any errors. To put it in pseudocode similar to your examples:

```
try {
    await outerBeforEach();
} catch (e) {
   reportError(e);
}
innerBeforeEach();
```


You can see this by patching Jasmine to log the timeout to the console (look for “Async function did not complete”):

root: begin
Timeout - Async function did not complete within 5000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
inner: begin
root: end
inner: end

​As you can see, the inner `beforeEach` doesn’t run until after the timeout.

The difference between Jasmine and most other testing frameworks is that it allows a spec to produce multiple errors. That’s a useful productivity boost when doing TDD and testing behaviors that can’t be fully expressed with a single assertion. It’s not what everyone wants, though, which is why we have the `oneFailurePerSpec` option. I could see a case for treating async timeouts in `beforeEach` or `beforeAll` as a special case and always handling them as if the `oneFailurePerSpec` option was set. I’d be happy to review a pull request that makes that change, although I don’t think it could be released until the next major release.

**Comment by sgravrock** [comment]
This was fixed by #1533 in 4.0.0.
